proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,483 @@
+
+* ``s"a ${let x =} b"``
+
+  * Incomplete expression
+
+* ``s"a ${{b} c"``
+
+  * The second ``{`` is not a valid character to start an expression
+
+See the "Semantics" and "Examples" sections to see examples of valid interpolated strings.
+
+Machinery
+~~~~~~~~~
+
+An interpolated string expression desugars to calls to ``fromBuilder``, ``toBuilder``, and ``interpolate``, which are defined as:
+
+::
+
+  -- | Laws:
+  --     * fromBuilder . toBuilder === id
+  --     * toBuilder . fromBuilder === id
+  class Monoid (Builder s) => Buildable s where
+    type Builder s = r | r -> s
+    toBuilder :: s -> Builder s
+    fromBuilder :: Builder s -> s
+
+  class Buildable s => Interpolate a s where
+    interpolate :: a -> Builder s
+
+Instances for ``String`` will be defined as well:
+
+::
+
+  newtype StringBuilder = StringBuilder (Endo String)
+    deriving newtype (Semigroup, Monoid)
+
+  instance Buildable String where
+    type Builder String = StringBuilder
+    toBuilder s = StringBuilder (Endo (s ++))
+    fromBuilder (StringBuilder (Endo f)) = f []
+
+  instance Interpolate String String where
+    interpolate = toBuilder
+  instance {-# OVERLAPPABLE #-} Show a => Interpolate a String where
+    interpolate = StringBuilder . Endo . shows
+
+This design allows interpolating anything, even user-defined types, into a ``String`` with ``Show``, but can be overridden for specific types. See the "Examples" section for more details.
+
+These definitions would initially be implemented in ``ghc-experimental`` under ``Data.String.Interpolate``. After the API has stablized, these would eventually live in ``GHC.Exts`` alongside ``IsString``.
+
+Semantics
+~~~~~~~~~
+
+With the machinery defined above, the following interpolated string desugars to the below expression:
+
+::
+
+  -- original string
+  s"foo ${f a b} bar ${g x} baz ${name}"
+
+  -- desugared
+  fromBuilder $
+    toBuilder "foo "
+    <> interpolate (f a b) <> toBuilder " bar "
+    <> interpolate (g x)   <> toBuilder " baz "
+    <> interpolate name    <> toBuilder ""
+
+The string literals there will be handled by ``-XOverloadedStrings`` as usual, if enabled.
+
+Interpolated multiline strings should desugar into a single-line string `as usual <https://github.com/brandonchinn178/ghc-proposals/blob/master/proposals/0569-multiline-strings.rst>`_ first, before interpolating expressions.
+
+::
+
+  let x = "hello"
+
+  -- original string
+  let str0 =
+        s"""
+        ${x} world
+        world ${x}
+        ${x} world
+        """
+
+  -- resolve multiline string
+  let str1 = s"${x} world\nworld ${x}\n${x} world"
+
+  -- resolve interpolation
+  let str2 = "hello world\nworld hello\nhello world"
+
+Examples
+--------
+
+Examples were tested with `this gist <https://gist.github.com/brandonchinn178/4d35ed189d7018ca34535ac85442790b>`_ (after desugaring the string interpolation).
+
+Edge cases
+~~~~~~~~~~
+
+The following interpolated string expressions are also valid:
```

Comment:
It used to be the case that an unmatched `}` would immediately close the `${}`. Is that still the case? Worth being explicit about.

I'd suggest including the following edge cases (I'm not sure if they're valid):

```
s"foo ${'}'} bar" (does it try to lex `'` by itself?)
s"foo ${'{'} bar" (because the `}` is matched, does it keep looking for the closing `}` until it gets to the end?)
s"foo ${ expr -- line comment } bar"
s"foo ${ expr {- inline comment -} } bar"

Reply:
Thanks, I'll add some more examples.

No, it's no longer the case that an unmatched `}` immediately closes. Instead, an expression is parsed as much as it can, then if there's no more expression, we expect a closing `}`.